{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a6e7e77f-e456-4782-b5da-90d985bea108",
   "metadata": {},
   "source": [
    "# Find the minimal medium #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fbbdfd18-5ede-4f02-9c4a-0d6344cc5b8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-17 16:18:20,157 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:18:20,160 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:18:20,161 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:18:20,163 - pycomo - INFO - Multiprocess Logger initialized.\n"
     ]
    }
   ],
   "source": [
    "import pycomo\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fbd1c942-4063-4a0b-8cd0-a33dbda8ec62",
   "metadata": {},
   "outputs": [],
   "source": [
    "pycomo.configure_logger(level=\"warning\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8781c9b-6199-45c2-831b-4b69ed47afbe",
   "metadata": {},
   "source": [
    "## Create the community model ##"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a3958ba-6ac0-455a-9eea-9edb3478e45b",
   "metadata": {},
   "source": [
    "First, we load the single organism models of the members. The models were automatically reconstructed using gapseq (see [Zimmermann et al. 2021](https://doi.org/10.1186/s13059-021-02295-1)), based on genome sequencing data:\n",
    " - M. smithii: [ABYW00000000.1](https://www.ncbi.nlm.nih.gov/nuccore/ABYW00000000.1/)\n",
    " - B. thetaiotaomicron [NC_004663.1](https://www.ncbi.nlm.nih.gov/nuccore/NC_004663.1)\n",
    "\n",
    "A full description of the model reconstruction can be found in [Duller et al. 2024](https://doi.org/10.1038/s41467-024-52037-7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "046498de-42ba-4920-9a8f-6281e5c6d544",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_models = \"data/member_models\"\n",
    "named_models = pycomo.load_named_models_from_dir(single_models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2f7af2f-abee-4a49-af5b-d9cea67f99f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ABYW00000000.1_ms_medium_gapseq_open': <Model ABYW00000000_1 at 0x7f7d90b8a150>,\n",
       " 'NC_004663.1_ms_b_tim_medium_gapseq_open': <Model NC_004663_1 at 0x7f7d9020df50>}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "named_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "696b65fe-afc8-4e9c-85c7-a35b68931743",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-17 16:18:27,103 - pycomo - WARNING - Warning: model name ABYW00000000.1_ms_medium_gapseq_open is not compliant with sbml id standards and was changed to ABYW00000000_1_ms_medium_gapseq_open\n",
      "2025-11-17 16:18:27,338 - pycomo - WARNING - Warning: model name NC_004663.1_ms_b_tim_medium_gapseq_open is not compliant with sbml id standards and was changed to NC_004663_1_ms_b_tim_medium_gapseq_open\n"
     ]
    }
   ],
   "source": [
    "prepared_models = [pycomo.SingleOrganismModel(m, n) for n, m in named_models.items()]\n",
    "com_model = pycomo.CommunityModel(prepared_models, name=\"com\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83e25e38-d7fa-429b-96e4-773488435c06",
   "metadata": {},
   "source": [
    "A report of the community model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "89ff1d89-8e1c-46ba-814e-fb63afe3135d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-17 16:19:00,003 - pycomo - WARNING - Not all reactions in the model are mass and charge balanced. To check which reactions are imbalanced, please run the get_unbalanced_reactions method of this CommunityModel object\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reactions: 9290\n",
      "Metabolites: 8786\n"
     ]
    }
   ],
   "source": [
    "print(f\"Reactions: {len(com_model.model.reactions)}\")\n",
    "print(f\"Metabolites: {len(com_model.model.metabolites)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b2d45e3-a6e4-4800-9913-a0d4b7ec2b5c",
   "metadata": {},
   "source": [
    "## Load and apply medium ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2683acb-e384-4065-a077-ac8f13c02bb4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>com</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7f7d8cd313d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>8786</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>9290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1141</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>987</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*community_biomass - 1.0*community_biomass_reverse_44dc1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>ABYW00000000_1_ms_medium_gapseq_open_c0, ABYW00000000_1_ms_medium_gapseq_open_e0, ABYW00000000_1_ms_medium_gapseq_open_p0, ABYW00000000_1_ms_medium_gapseq_open_bio, ABYW00000000_1_ms_medium_gapseq_open_medium, medium, fraction_reaction, NC_004663_1_ms_b_tim_medium_gapseq_open_c0, NC_004663_1_ms_b_tim_medium_gapseq_open_e0, NC_004663_1_ms_b_tim_medium_gapseq_open_p0, NC_004663_1_ms_b_tim_medium_gapseq_open_bio, NC_004663_1_ms_b_tim_medium_gapseq_open_medium</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model com at 0x7f7d8cd313d0>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medium_dict = pycomo.read_medium_from_file(\"data/media/ms_b_tim_medium_gapseq_open.csv\", \"_medium\")\n",
    "for k, v in medium_dict.items():\n",
    "    medium_dict[k] = 1000.\n",
    "com_model.medium = medium_dict\n",
    "com_model.apply_medium()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0381b7f8-a350-4fd2-af3f-5b616127e3cf",
   "metadata": {},
   "source": [
    "## Save the model ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "11ed4d87-3f0d-42a6-9529-bd0390e8dd38",
   "metadata": {},
   "outputs": [],
   "source": [
    "com_model.save(\"data/community_models/M_smithii_B_tim_ms_b_tim_medium_gapseq_open_unconstr.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "df0e94e4-c034-445c-a2b1-be908d7ba37b",
   "metadata": {},
   "outputs": [],
   "source": [
    "com_model = pycomo.CommunityModel.load(\"data/community_models/M_smithii_B_tim_ms_b_tim_medium_gapseq_open_unconstr.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71ff6fb9-c1fc-4dbf-abb4-e30c4a3437b8",
   "metadata": {},
   "source": [
    "## Set the transport reaction bounds to physiologically feasible range ##\n",
    "Since data for transport fluxes is not present, we set transport bounds to $\\pm 10 mmol * gDW^{-1} * h^{-1}$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10afe37c-3f61-4ec7-8d8c-9c6e576570d9",
   "metadata": {},
   "source": [
    "def is_transporter(rxn, member_names):\n",
    "    if \"_TF_\" in rxn.id or rxn.id[-3:] in [\"_lb\", \"_ub\"]:\n",
    "        return False\n",
    "    met_comps = [met.compartment for met in rxn.metabolites.keys() if met.compartment != \"fraction_reaction\"]\n",
    "    if len(set(met_comps)) < 2:\n",
    "        return False\n",
    "    for n in member_names:\n",
    "        if n in rxn.id and all([n in m for m in met_comps]):\n",
    "            return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "45b6aca8-0bed-41e7-8cbb-617fd5d66cea",
   "metadata": {},
   "outputs": [],
   "source": [
    "member_names = com_model.get_member_names()\n",
    "name_conv = {\"A\": \"M. smithii\", \"N\": \"B. thetaiotaomicron\"}\n",
    "\n",
    "member_abbrev = [name_conv[n[0]] for n in member_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "59c15bd2-181a-409f-87cd-5dd397e25288",
   "metadata": {},
   "outputs": [],
   "source": [
    "tp_val = 10.\n",
    "for tp_rxn in com_model.model.reactions.query(lambda r: com_model.is_transporter(r)):\n",
    "    lb = -tp_val if tp_rxn.lower_bound < -tp_val else tp_rxn.lower_bound\n",
    "    ub = tp_val if tp_rxn.upper_bound > tp_val else tp_rxn.upper_bound\n",
    "    com_model.change_reaction_bounds(tp_rxn, lower_bound=lb, upper_bound=ub)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ef44782-ff9e-403f-b7f8-3ae926037cc4",
   "metadata": {},
   "source": [
    "## Find a minimal medium for the community, based on the MS medium ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5d9a5797-b77d-41dc-be48-ecb4a71b8287",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_cpd00009_medium    0.098943\n",
      "EX_cpd00030_medium    0.000365\n",
      "EX_cpd00034_medium    0.000365\n",
      "EX_cpd00051_medium    0.134785\n",
      "EX_cpd00119_medium    0.210719\n",
      "EX_cpd00149_medium    0.000585\n",
      "EX_cpd00244_medium    0.000094\n",
      "EX_cpd10516_medium    0.000365\n",
      "EX_cpd00084_medium    0.007729\n",
      "EX_cpd00063_medium    0.000365\n",
      "EX_cpd00058_medium    0.000365\n",
      "EX_cpd10515_medium    0.000759\n",
      "EX_cpd00060_medium    0.017689\n",
      "EX_cpd00205_medium    0.000365\n",
      "EX_cpd00254_medium    0.000365\n",
      "EX_cpd00393_medium    0.000597\n",
      "EX_cpd00220_medium    0.000395\n",
      "EX_cpd00048_medium    0.001331\n",
      "EX_cpd00099_medium    0.000365\n",
      "EX_cpd90020_medium    0.002532\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "com_model.convert_to_fixed_abundance()\n",
    "min_med = cobra.medium.minimal_medium(com_model.model)\n",
    "print(min_med)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fa8e9366-9585-444a-be84-881833b42338",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compounds</th>\n",
       "      <th>name</th>\n",
       "      <th>maxFlux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cpd00009</td>\n",
       "      <td>Phosphate</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cpd00030</td>\n",
       "      <td>Mn2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cpd00034</td>\n",
       "      <td>Zn2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cpd00051</td>\n",
       "      <td>L-Arginine</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cpd00119</td>\n",
       "      <td>L-Histidine</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cpd00149</td>\n",
       "      <td>Co2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cpd00244</td>\n",
       "      <td>Ni2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cpd10516</td>\n",
       "      <td>fe3</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cpd00084</td>\n",
       "      <td>L-Cysteine</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cpd00063</td>\n",
       "      <td>Ca2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cpd00058</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cpd10515</td>\n",
       "      <td>Fe2+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cpd00060</td>\n",
       "      <td>L-Methionine</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cpd00205</td>\n",
       "      <td>K+</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cpd00254</td>\n",
       "      <td>Mg</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cpd00393</td>\n",
       "      <td>Folate</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cpd00220</td>\n",
       "      <td>Riboflavin</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cpd00048</td>\n",
       "      <td>Sulfate</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cpd00099</td>\n",
       "      <td>Cl-</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>cpd90020</td>\n",
       "      <td>Cellulose</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   compounds          name  maxFlux\n",
       "0   cpd00009     Phosphate     1000\n",
       "1   cpd00030          Mn2+     1000\n",
       "2   cpd00034          Zn2+     1000\n",
       "3   cpd00051    L-Arginine     1000\n",
       "4   cpd00119   L-Histidine     1000\n",
       "5   cpd00149          Co2+     1000\n",
       "6   cpd00244          Ni2+     1000\n",
       "7   cpd10516           fe3     1000\n",
       "8   cpd00084    L-Cysteine     1000\n",
       "9   cpd00063          Ca2+     1000\n",
       "10  cpd00058          Cu2+     1000\n",
       "11  cpd10515          Fe2+     1000\n",
       "12  cpd00060  L-Methionine     1000\n",
       "13  cpd00205            K+     1000\n",
       "14  cpd00254            Mg     1000\n",
       "15  cpd00393        Folate     1000\n",
       "16  cpd00220    Riboflavin     1000\n",
       "17  cpd00048       Sulfate     1000\n",
       "18  cpd00099           Cl-     1000\n",
       "19  cpd90020     Cellulose     1000"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medium_dict = min_med.to_dict()\n",
    "\n",
    "medium_df_rows = []\n",
    "\n",
    "for exchg_rxn in medium_dict.keys():\n",
    "    met = list(com_model.model.reactions.get_by_id(exchg_rxn).metabolites.keys())[0]\n",
    "    met_id = met.id.replace(\"_medium\",\"\")\n",
    "    met_name = met.name.replace(\"_medium\", \"\")\n",
    "    medium_df_rows.append({\"compounds\": met_id, \"name\": met_name, \"maxFlux\": 1000})\n",
    "\n",
    "medium_df = pd.DataFrame.from_dict(medium_df_rows)\n",
    "\n",
    "medium_df.to_csv(\"data/media/minimal_medium_community.csv\", index=False, sep=\",\")\n",
    "\n",
    "medium_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59a08678-7c30-416b-9fa2-9c15b05d0228",
   "metadata": {},
   "source": [
    "## Apply the minimal medium to the model ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f91921c1-55fe-44bd-8258-d03f5050b5e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>com</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>7f7d85a4dad0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>9008</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>9512</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>1141</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>987</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*community_biomass - 1.0*community_biomass_reverse_44dc1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>ABYW00000000_1_ms_medium_gapseq_open_c0, ABYW00000000_1_ms_medium_gapseq_open_e0, ABYW00000000_1_ms_medium_gapseq_open_p0, ABYW00000000_1_ms_medium_gapseq_open_bio, ABYW00000000_1_ms_medium_gapseq_open_medium, medium, fraction_reaction, NC_004663_1_ms_b_tim_medium_gapseq_open_c0, NC_004663_1_ms_b_tim_medium_gapseq_open_e0, NC_004663_1_ms_b_tim_medium_gapseq_open_p0, NC_004663_1_ms_b_tim_medium_gapseq_open_bio, NC_004663_1_ms_b_tim_medium_gapseq_open_medium</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model com at 0x7f7d85a4dad0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medium_dict = min_med.to_dict()\n",
    "for k, v in medium_dict.items():\n",
    "    medium_dict[k] = 1000.\n",
    "com_model.medium = medium_dict\n",
    "com_model.apply_medium()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1df4e8c4-c0ea-464b-959c-3ae5b015a56c",
   "metadata": {},
   "source": [
    "## Save the model ##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e4c44b5e-f35c-4639-9d46-46792747affe",
   "metadata": {},
   "outputs": [],
   "source": [
    "com_model.save(f\"data/community_models/{member_abbrev[0]}_{member_abbrev[1]}_ms_open_min_med_tp_10.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca72666d-71d4-4354-b7d8-0379aa1a553b",
   "metadata": {},
   "source": [
    "## Find the maximum growth rate ##"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a8884fc-5687-46e6-aeb7-efc7294df6fd",
   "metadata": {},
   "source": [
    "_M. smithii_ growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e2cab67e-b843-4188-a14e-cbe68a40fbdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ABYW00000000_1_ms_medium_gapseq_open',\n",
       " 'NC_004663_1_ms_b_tim_medium_gapseq_open']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "member_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "92b797e9-918c-4316-9269-4f1d20055eb8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Growth rate of M. smithii: 0.0 (solver status: optimal)\n"
     ]
    }
   ],
   "source": [
    "com_model.convert_to_fixed_abundance()\n",
    "com_model.apply_fixed_abundance({member_names[0]: 1., member_names[1]: 0.})\n",
    "solution = com_model.model.optimize()\n",
    "print(f\"Growth rate of M. smithii: {solution.objective_value} (solver status: {solution.status})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a8a15ed-13d3-40a9-a49a-5aa115fda93a",
   "metadata": {},
   "source": [
    "_B. thetaiotaomicron_ growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6fa8698f-f269-44d9-9c6a-d30c62cd5ce5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Growth rate of B. thetaiotaomicron: 0.7754058387708818 (solver status: optimal)\n"
     ]
    }
   ],
   "source": [
    "com_model.convert_to_fixed_abundance()\n",
    "com_model.apply_fixed_abundance({member_names[0]: 0., member_names[1]: 1.})\n",
    "solution = com_model.model.optimize()\n",
    "print(f\"Growth rate of B. thetaiotaomicron: {solution.objective_value} (solver status: {solution.status})\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15303098-4458-4ba2-bfdc-94f28b29d0ec",
   "metadata": {},
   "source": [
    "Community growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8fb104cf-8617-4870-903e-f2884f92a29b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-17 16:22:56,003 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:22:56,005 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:22:56,005 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:22:56,006 - pycomo - INFO - Multiprocess Logger initialized.\n",
      "2025-11-17 16:22:57,940 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:22:57,941 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:22:57,941 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:22:57,941 - pycomo - INFO - Multiprocess Logger initialized.\n",
      "2025-11-17 16:23:24,624 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:23:24,624 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:23:24,624 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:23:24,625 - pycomo - INFO - Multiprocess Logger initialized.\n",
      "2025-11-17 16:23:26,708 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:23:26,708 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:23:26,709 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:23:26,709 - pycomo - INFO - Multiprocess Logger initialized.\n",
      "2025-11-17 16:23:39,934 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:23:39,935 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:23:39,935 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:23:39,936 - pycomo - INFO - Multiprocess Logger initialized.\n",
      "2025-11-17 16:23:41,824 - pycomo - INFO - Logger initialized.\n",
      "2025-11-17 16:23:41,824 - pycomo - INFO - Process Pool Logger initialized.\n",
      "2025-11-17 16:23:41,825 - pycomo - INFO - Utils Logger initialized.\n",
      "2025-11-17 16:23:41,825 - pycomo - INFO - Multiprocess Logger initialized.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>min_flux</th>\n",
       "      <th>max_flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABYW00000000_1_ms_medium_gapseq_open_fraction_...</td>\n",
       "      <td>0.478045</td>\n",
       "      <td>0.479571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_004663_1_ms_b_tim_medium_gapseq_open_fracti...</td>\n",
       "      <td>0.520429</td>\n",
       "      <td>0.521955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>community_biomass</td>\n",
       "      <td>0.851180</td>\n",
       "      <td>0.851180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         reaction_id  min_flux  max_flux\n",
       "0  ABYW00000000_1_ms_medium_gapseq_open_fraction_...  0.478045  0.479571\n",
       "1  NC_004663_1_ms_b_tim_medium_gapseq_open_fracti...  0.520429  0.521955\n",
       "2                                  community_biomass  0.851180  0.851180"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "com_model.max_growth_rate(sensitivity=5, return_abundances=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "278a2ed5-a62e-46f7-ace2-f059452ba11c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pycomo_py_3.11",
   "language": "python",
   "name": "pycomo_py_3.11"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
